File: AoC_03_Lobby.zig
const std = @import("std");

pub fn main() !void {
    try part_one();
    try part_two();
}

pub fn part_one() !void {
    const file = try std.fs.cwd().openFile("input.txt", .{});
    defer file.close();

    var buffer: [1000]u8 = undefined;
    var file_reader = file.reader(&buffer);
    var rating: i64 = 0;

    while (file_reader.interface.takeDelimiterExclusive('\n')) |line| {
        rating += max_rating(line, 2);
        if (!file_reader.atEnd())
            file_reader.interface.toss(1);
    } else |_| {}
    std.debug.print("{d}\n", .{rating});
}

pub fn part_two() !void {
    const file = try std.fs.cwd().openFile("input.txt", .{});
    defer file.close();

    var buffer: [1000]u8 = undefined;
    var file_reader = file.reader(&buffer);
    var rating: i64 = 0;

    while (file_reader.interface.takeDelimiterExclusive('\n')) |line| {
        rating += max_rating(line, 12);
        if (!file_reader.atEnd())
            file_reader.interface.toss(1);
    } else |_| {}
    std.debug.print("{d}\n", .{rating});
}

fn max_rating(line: []u8, bank: comptime_int) i64 {
    var i: usize = 0;
    var k: usize = 0;
    var max: usize = 0;
    var rating: [bank]u8 = undefined;
    while (k < bank) : (k += 1) {
        i = max;
        while (i < line.len - bank + k + 1) : (i += 1) {
            if (line[i] > line[max]) {
                max = i;
            }
        }
        rating[k] = line[max];
        max += 1;
    }
    return std.fmt.parseInt(i64, rating[0..bank], 10) catch 0;
}

File: AoC_03_Lobby.cs
var data = File.ReadAllLines(@"input.txt");
PartOne();
PartTwo();

void PartOne()
{
    long result = 0;
    foreach (var line in data)
    {
        result += MaxRating(line, 2);
    }
    Console.WriteLine(result);
}

void PartTwo()
{
    long result = 0;
    foreach (var line in data)
    {
        result += MaxRating(line, 12);
    }
    Console.WriteLine(result);
}

long MaxRating(string line, int bank)
{
    int i = 0, k, max = 0;
    char[] rating = new char[bank];

    for (k = 0; k < bank; ++k)
    {
        for (i = max; i < line.Length - bank + k + 1; ++i)
        {
            if (line[i] > line[max])
            {
                max = i;
            }
        }
        rating[k] = line[max];
        max++;
    }

    return long.Parse(new string(rating));
}